File: server/lib/agent/planner.dart
import 'dart:convert';

import '../models/work_item.dart';
import '../ollama/ollama_client.dart';

class PlannedStep {
  PlannedStep({required this.title, required this.rationale});
  final String title;
  final String? rationale;
}

class Planner {
  Planner(this._ollama);
  final OllamaClient _ollama;

  static const _systemPrompt = '''
You are the planning module of an agentic DevOps assistant. Given a work
item, produce a SHORT ordered plan of investigative steps a software
engineer should take to triage and resolve it. Steps should be concrete
and verifiable; prefer 3-6 steps.

Respond with STRICT JSON in this shape:
{
  "steps": [
    {"title": "<short imperative title>", "rationale": "<why this step>"}
  ]
}
Do not include any prose outside the JSON object.
''';

  Future<List<PlannedStep>> plan(WorkItem item) async {
    final response = await _ollama.chat(
      messages: [
        OllamaMessage(role: 'system', content: _systemPrompt),
        OllamaMessage(role: 'user', content: _renderItem(item)),
      ],
      format: {'type': 'object'},
      temperature: 0.1,
    );
    return _parse(response.content);
  }

  String _renderItem(WorkItem i) {
    final buf = StringBuffer()
      ..writeln('Work item type: ${i.type}')
      ..writeln('Title: ${i.title}')
      ..writeln('Severity: ${i.severity}  Priority: ${i.priority}')
      ..writeln('State: ${i.state}')
      ..writeln('Area: ${i.areaPath}')
      ..writeln('Tags: ${i.tags ?? "(none)"}');
    if (i.description != null && i.description!.trim().isNotEmpty) {
      buf
        ..writeln()
        ..writeln('Description:')
        ..writeln(i.description);
    }
    if (i.reproSteps != null && i.reproSteps!.trim().isNotEmpty) {
      buf
        ..writeln()
        ..writeln('Repro steps:')
        ..writeln(i.reproSteps);
    }
    if (i.systemInfo != null && i.systemInfo!.trim().isNotEmpty) {
      buf
        ..writeln()
        ..writeln('System info:')
        ..writeln(i.systemInfo);
    }
    return buf.toString();
  }

  List<PlannedStep> _parse(String raw) {
    final cleaned = _extractJson(raw);
    final decoded = jsonDecode(cleaned);
    if (decoded is! Map || decoded['steps'] is! List) {
      throw FormatException('planner response missing "steps" list: $raw');
    }
    final steps = (decoded['steps'] as List).cast<Object?>();
    return steps.map((s) {
      final m = (s as Map).cast<String, Object?>();
      return PlannedStep(
        title: (m['title'] as String?)?.trim() ?? 'untitled step',
        rationale: (m['rationale'] as String?)?.trim(),
      );
    }).toList(growable: false);
  }

  /// Models sometimes wrap JSON in markdown fences; strip them.
  String _extractJson(String raw) {
    final trimmed = raw.trim();
    if (trimmed.startsWith('```')) {
      final firstNewline = trimmed.indexOf('\n');
      final body = firstNewline == -1 ? trimmed : trimmed.substring(firstNewline + 1);
      final endFence = body.lastIndexOf('```');
      return (endFence == -1 ? body : body.substring(0, endFence)).trim();
    }
    return trimmed;
  }
}

File: server/app/agent/planner.py
from __future__ import annotations

import json
from dataclasses import dataclass

from ..models.work_item import WorkItem
from .ollama_client import OllamaClient, OllamaMessage

_SYSTEM = """You are the planning module of an agentic DevOps assistant. Given a work
item, produce a SHORT ordered plan of investigative steps a software engineer
should take to triage and resolve it. Steps should be concrete and verifiable;
prefer 3-6 steps.

Respond with STRICT JSON in this shape:
{
  "steps": [
    {"title": "<short imperative title>", "rationale": "<why this step>"}
  ]
}
Do not include any prose outside the JSON object.
"""


@dataclass
class PlannedStep:
    title: str
    rationale: str | None


class Planner:
    def __init__(self, ollama: OllamaClient) -> None:
        self._ollama = ollama

    async def plan(self, item: WorkItem) -> list[PlannedStep]:
        response = await self._ollama.chat(
            messages=[
                OllamaMessage(role="system", content=_SYSTEM),
                OllamaMessage(role="user", content=self._render(item)),
            ],
            response_format={"type": "object"},
            temperature=0.1,
        )
        return self._parse(response.content)

    def _render(self, i: WorkItem) -> str:
        lines = [
            f"Work item type: {i.type}",
            f"Title: {i.title}",
            f"Severity: {i.severity}  Priority: {i.priority}",
            f"State: {i.state}",
            f"Area: {i.area_path}",
            f"Tags: {i.tags or '(none)'}",
        ]
        if i.description and i.description.strip():
            lines += ["", "Description:", i.description]
        if i.repro_steps and i.repro_steps.strip():
            lines += ["", "Repro steps:", i.repro_steps]
        if i.system_info and i.system_info.strip():
            lines += ["", "System info:", i.system_info]
        return "\n".join(lines)

    def _parse(self, raw: str) -> list[PlannedStep]:
        cleaned = _strip_fences(raw)
        decoded = json.loads(cleaned)
        if not isinstance(decoded, dict) or not isinstance(decoded.get("steps"), list):
            raise ValueError(f'planner response missing "steps" list: {raw!r}')
        out: list[PlannedStep] = []
        for s in decoded["steps"]:
            if not isinstance(s, dict):
                continue
            out.append(
                PlannedStep(
                    title=str(s.get("title", "untitled step")).strip(),
                    rationale=(s.get("rationale") or None),
                )
            )
        return out


def _strip_fences(raw: str) -> str:
    t = raw.strip()
    if not t.startswith("```"):
        return t
    first_nl = t.find("\n")
    body = t[first_nl + 1 :] if first_nl != -1 else t
    end = body.rfind("```")
    return (body[:end] if end != -1 else body).strip()
